File: native/android/core/src/main/kotlin/dev/respondkit/core/RespondKitStore.kt
package dev.respondkit.core

import java.time.Instant
import kotlinx.coroutines.*
import kotlinx.coroutines.flow.*
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json

/**
 * Host-owned, main-thread-confined store. Retain above the screen; cancel its scope on disposal.
 */
class RespondKitStore(
    private val configuration: RespondKitConfiguration,
    private var context: CustomerContext = CustomerContext(),
    private val persistence: RespondKitPersistence,
    private val api: RespondKitApi = RespondKitClient(configuration),
    private val scope: CoroutineScope,
    private var identityToken: (suspend () -> String)? = null,
) {
    private val json = Json { ignoreUnknownKeys = true }
    private var stored =
        persistence.load()?.let { json.decodeFromString<StoredState>(it) }
            ?: StoredState(userId = context.userId)
    private val mutableState = MutableStateFlow(SupportState())
    val state: StateFlow<SupportState> = mutableState.asStateFlow()
    private var session: ClientSession? = null
    private var epoch = 0L
    private var pollJob: Job? = null
    private var screenVisible = false
    private val mutex = Mutex()

    init {
        if (stored.userId != context.userId) stored = StoredState(userId = context.userId)
        stored =
            stored.copy(
                pending =
                    stored.pending.mapValues { (_, messages) ->
                        messages.map {
                            if (it.delivery == "sending") it.copy(delivery = "acceptance_unknown")
                            else it
                        }
                    }
            )
        persist()
        publish()
    }

    fun setForeground(active: Boolean) {
        if (state.value.isForeground == active) return
        mutableState.update { it.copy(isForeground = active) }
        pollJob?.cancel()
        pollJob = null
        if (active)
            pollJob =
                scope.launch {
                    while (isActive) {
                        refresh()
                        delay(configuration.pollIntervalMillis)
                    }
                }
    }

    fun setScreenVisible(visible: Boolean) {
        screenVisible = visible
    }

    fun selectThread(id: String?) {
        mutableState.update { it.copy(activeThreadId = id) }
        publish()
    }

    fun setDraft(text: String) {
        stored =
            stored.copy(drafts = stored.drafts + ((state.value.activeThreadId ?: "new") to text))
        persistOrReport()
        publish()
    }

    fun clearError() {
        mutableState.update { it.copy(errorMessage = null) }
    }

    /**
     * Call after host auth settles. Account changes isolate UI synchronously, before network
     * revocation.
     */
    fun updateIdentity(context: CustomerContext, identityToken: (suspend () -> String)? = null) {
        epoch++
        val previous = session
        val changed = this.context.userId != context.userId
        this.context = context
        this.identityToken = identityToken
        session = null
        if (changed) {
            stored = StoredState(userId = context.userId)
            mutableState.update { it.copy(activeThreadId = null) }
            publish()
            persist()
            if (previous != null)
                scope.launch {
                    try {
                        api.logout(previous.token)
                    } catch (error: CancellationException) {
                        throw error
                    } catch (error: Exception) {
                        mutableState.update {
                            it.copy(
                                errorMessage =
                                    "Previous support session could not be revoked. It will expire automatically."
                            )
                        }
                    }
                }
        }
        publish()
    }

    suspend fun refresh() = operate { generation ->
        val statuses = mutableListOf<ThreadStatus>()
        var after: String? = null
        val seen = mutableSetOf<String>()
        do {
            val page = authorized(generation) { api.statuses(it, after) }
            check(generation)
            page.threads.forEach { replyCursor(it.latestReplyCursor) }
            statuses += page.threads
            after = page.nextCursor
            if (after != null && !seen.add(after))
                throw RespondKitException("Repeated history cursor.")
        } while (after != null)
        stored = stored.copy(statuses = statuses.sortedByDescending { it.thread.updatedAt })
        persist()
        publish()
        state.value.activeThreadId?.takeIf { screenVisible }?.let { loadMessages(it, generation) }
        flushReads(generation)
    }

    suspend fun sendDraft() {
        val text = state.value.draft
        val id = state.value.activeThreadId
        if (text.isBlank() || text.length > 6_000) {
            mutableState.update {
                it.copy(errorMessage = "Enter a message of up to 6,000 characters.")
            }
            return
        }
        operate { generation ->
            if (
                id != null &&
                    stored.statuses.firstOrNull { it.thread.id == id }?.thread?.state == "closed"
            )
                throw RespondKitException("This conversation is closed. Start a new conversation.")
            val key = id ?: "new"
            val pending = PendingMessage(newId("cmsg"), text, now(), "sending")
            stored =
                stored.copy(
                    pending = stored.pending + (key to (stored.pending[key].orEmpty() + pending)),
                    drafts =
                        if (stored.drafts[key] == text) stored.drafts + (key to "")
                        else stored.drafts,
                )
            persist()
            publish()
            deliver(pending, id, generation)
        }
    }

    suspend fun retry(messageId: String) {
        val id = state.value.activeThreadId
        operate { generation ->
            stored.pending[id ?: "new"]
                ?.firstOrNull { it.id == messageId }
                ?.let { deliver(it, id, generation) }
        }
    }

    suspend fun markDisplayed(threadId: String, cursor: String) {
        val current = state.value
        if (
            !current.isForeground ||
                !screenVisible ||
                current.activeThreadId != threadId ||
                current.loadedCursor != cursor ||
                current.messages.none { it.isReply }
        )
            return
        operate { generation ->
            val latest = state.value
            if (
                !latest.isForeground ||
                    !screenVisible ||
                    latest.activeThreadId != threadId ||
                    latest.loadedCursor != cursor
            )
                return@operate
            if (replyCursor(cursor) <= replyCursor(stored.readCursors[threadId] ?: "0"))
                return@operate
            stored =
                stored.copy(
                    readCursors = stored.readCursors + (threadId to cursor),
                    pendingReads = stored.pendingReads + (threadId to cursor),
                )
            persist()
            publish()
            flushReads(generation)
        }
    }

    private suspend fun deliver(pending: PendingMessage, threadId: String?, generation: Long) {
        var id = threadId
        try {
            if (id == null) {
                val created =
                    authorized(generation) { api.createThread(it, stored.newClientThreadId) }
                check(generation)
                id = created.id
                stored =
                    stored.copy(
                        statuses =
                            listOf(ThreadStatus(created, "0")) +
                                stored.statuses.filterNot { it.thread.id == created.id },
                        pending =
                            (stored.pending - "new") +
                                (created.id to
                                    (stored.pending[created.id].orEmpty() +
                                        stored.pending["new"].orEmpty())),
                        drafts =
                            (stored.drafts - "new") +
                                (created.id to stored.drafts["new"].orEmpty()),
                        newClientThreadId = newId("cthread"),
                    )
                mutableState.update { it.copy(activeThreadId = created.id) }
                persist()
                publish()
            }
            val actualId = id
            setDelivery(pending.id, actualId, "sending")
            persist()
            publish()
            val accepted =
                authorized(generation) { api.send(it, actualId, pending.id, pending.text) }
            check(generation)
            setDelivery(
                pending.id,
                actualId,
                when (accepted.status) {
                    "failed" -> "failed"
                    "acceptance_unknown" -> "acceptance_unknown"
                    else -> "accepted"
                },
            )
            persist()
            publish()
            loadMessages(actualId, generation)
        } catch (error: Exception) {
            if (generation != epoch) throw error
            setDelivery(pending.id, id ?: "new", "acceptance_unknown")
            persist()
            publish()
            throw error
        }
    }

    private suspend fun loadMessages(id: String, generation: Long) {
        var cursor = stored.cursors[id] ?: "0"
        val messages = stored.messages[id].orEmpty().associateBy { it.id }.toMutableMap()
        var more: Boolean
        do {
            val page = authorized(generation) { api.messages(it, id, cursor) }
            check(generation)
            val next = replyCursor(page.nextCursor)
            if (next < replyCursor(cursor) || (page.hasMore && next <= replyCursor(cursor)))
                throw RespondKitException("Invalid message pagination.")
            page.messages.forEach { messages[it.id] = it }
            cursor = page.nextCursor
            more = page.hasMore
        } while (more)
        val canonical =
            messages.values
                .mapNotNull { message -> message.clientMessageId?.let { it to message } }
                .toMap()
        val pending =
            stored.pending[id].orEmpty().mapNotNull { pending ->
                val message = canonical[pending.id]
                when {
                    message == null -> pending
                    message.state == "failed" -> pending.copy(delivery = "failed")
                    else -> null
                }
            }
        val failed =
            messages.values
                .filter {
                    it.state == "failed" &&
                        it.clientMessageId != null &&
                        pending.none { p -> p.id == it.clientMessageId }
                }
                .map { PendingMessage(it.clientMessageId!!, it.text, it.acceptedAt, "failed") }
        stored =
            stored.copy(
                messages = stored.messages + (id to messages.values.sortedBy { it.acceptedAt }),
                cursors = stored.cursors + (id to cursor),
                pending = stored.pending + (id to (pending + failed)),
            )
        persist()
        publish()
    }

    private suspend fun flushReads(generation: Long) {
        for ((id, cursor) in stored.pendingReads) {
            authorized(generation) { api.markRead(it, id, cursor) }
            check(generation)
            stored = stored.copy(pendingReads = stored.pendingReads - id)
            persist()
        }
    }

    private suspend fun validToken(generation: Long): String {
        persist()
        session
            ?.takeIf { Instant.parse(it.expiresAt).isAfter(Instant.now().plusSeconds(15)) }
            ?.let {
                return it.token
            }
        val assertion = identityToken?.invoke()
        check(generation)
        if (identityToken != null && assertion.isNullOrBlank())
            throw RespondKitException("Support identity could not be verified.")
        val created = api.createSession(stored.installationId, context, assertion)
        check(generation)
        session = created
        return created.token
    }

    private suspend fun <T> authorized(generation: Long, action: suspend (String) -> T): T =
        try {
            action(validToken(generation))
        } catch (error: RespondKitException) {
            if (error.status != 401) throw error
            check(generation)
            session = null
            action(validToken(generation))
        }

    private suspend fun check(expected: Long) {
        currentCoroutineContext().ensureActive()
        if (expected != epoch) throw CancellationException("Support identity changed")
    }

    private suspend fun operate(action: suspend (Long) -> Unit) {
        val generation = epoch
        mutex.withLock {
            if (generation != epoch) return
            mutableState.update { it.copy(isLoading = true) }
            try {
                action(generation)
                check(generation)
                clearError()
            } catch (error: CancellationException) {
                throw error
            } catch (error: Exception) {
                if (generation == epoch)
                    mutableState.update {
                        it.copy(errorMessage = error.message ?: "Support request failed.")
                    }
            } finally {
                mutableState.update { it.copy(isLoading = false) }
            }
        }
    }

    private fun setDelivery(messageId: String, threadId: String, value: String) {
        stored =
            stored.copy(
                pending =
                    stored.pending +
                        (threadId to
                            stored.pending[threadId].orEmpty().map {
                                if (it.id == messageId) it.copy(delivery = value) else it
                            })
            )
    }

    private fun persist() {
        persistence.save(json.encodeToString(stored))
    }

    private fun persistOrReport() {
        try {
            persist()
        } catch (error: Exception) {
            mutableState.update {
                it.copy(errorMessage = error.message ?: "Could not save support history.")
            }
        }
    }

    private fun publish() {
        val key = state.value.activeThreadId ?: "new"
        mutableState.update {
            it.copy(
                statuses = stored.statuses,
                unreadThreadIds =
                    stored.statuses
                        .filter { status ->
                            replyCursor(status.latestReplyCursor) >
                                replyCursor(stored.readCursors[status.thread.id] ?: "0")
                        }
                        .map { status -> status.thread.id }
                        .toSet(),
                messages = stored.messages[key].orEmpty(),
                pendingMessages = stored.pending[key].orEmpty(),
                draft = stored.drafts[key].orEmpty(),
                loadedCursor = stored.cursors[key] ?: "0",
            )
        }
    }
}

File: native/swift/Sources/RespondKitCore/RespondKitStore.swift
import Foundation
import Observation

/// Retain one store in the host app, independently of the presented screen.
@MainActor @Observable public final class RespondKitStore {
  public private(set) var statuses: [ThreadStatus] = []
  public private(set) var hasUnreadReplies = false
  public private(set) var activeThreadID: String?
  public private(set) var messages: [SupportMessage] = []
  public private(set) var pendingMessages: [PendingMessage] = []
  public private(set) var draft = ""
  public private(set) var loadedCursor = "0"
  public private(set) var isLoading = false
  public private(set) var errorMessage: String?
  public private(set) var isForeground = false
  public var activeThread: SupportThread? {
    statuses.first { $0.thread.id == activeThreadID }?.thread
  }

  @ObservationIgnored private let configuration: RespondKitConfiguration
  @ObservationIgnored private let api: any RespondKitAPI
  @ObservationIgnored private let persistence: any RespondKitPersistence
  @ObservationIgnored private var context: CustomerContext
  @ObservationIgnored private var identityToken: (@Sendable () async throws -> String)?
  @ObservationIgnored private var state: StoredState
  @ObservationIgnored private var session: ClientSession?
  @ObservationIgnored private var epoch = 0
  @ObservationIgnored private var pollTask: Task<Void, Never>?
  @ObservationIgnored private var screenVisible = false
  @ObservationIgnored private var locked = false
  @ObservationIgnored private var waiters: [CheckedContinuation<Void, Never>] = []

  public init(
    configuration: RespondKitConfiguration, context: CustomerContext = .init(),
    persistence: (any RespondKitPersistence)? = nil, api: (any RespondKitAPI)? = nil,
    identityToken: (@Sendable () async throws -> String)? = nil
  ) throws {
    self.configuration = configuration
    self.context = context
    self.identityToken = identityToken
    self.api = api ?? RespondKitClient(configuration: configuration)
    let storage = persistence ?? KeychainPersistence(scope: configuration.storageScope)
    self.persistence = storage
    var saved =
      try storage.load().map { try JSONDecoder().decode(StoredState.self, from: $0) }
      ?? StoredState(userID: context.userId)
    if saved.userID != context.userId { saved = StoredState(userID: context.userId) }
    // A process can terminate before acceptance arrives; retry the immutable pending payload.
    for key in saved.pending.keys {
      saved.pending[key] = saved.pending[key]?.map { message in
        var message = message
        if message.delivery == "sending" { message.delivery = "acceptance_unknown" }
        return message
      }
    }
    self.state = saved
    try storage.save(JSONEncoder().encode(saved))
    publish()
  }

  public func setForeground(_ active: Bool) {
    guard active != isForeground else { return }
    isForeground = active
    pollTask?.cancel()
    pollTask = nil
    guard active else { return }
    pollTask = Task { [weak self, interval = configuration.pollInterval] in
      while !Task.isCancelled {
        guard let self else { return }
        await self.refresh()
        do { try await Task.sleep(for: interval) } catch { return }
      }
    }
  }
  public func setScreenVisible(_ visible: Bool) { screenVisible = visible }
  public func selectThread(_ id: String?) {
    activeThreadID = id
    publish()
  }
  public func setDraft(_ text: String) {
    state.drafts[activeThreadID ?? "new"] = text
    persistOrReport()
    publish()
  }
  public func isUnread(_ threadID: String) -> Bool {
    guard let status = statuses.first(where: { $0.thread.id == threadID }) else { return false }
    return (Int64(status.latestReplyCursor) ?? 0) > (Int64(state.readCursors[threadID] ?? "0") ?? 0)
  }
  public func clearError() { errorMessage = nil }

  /// Call only after host authentication settles. A subject change isolates all cached state immediately.
  public func updateIdentity(
    context: CustomerContext, identityToken: (@Sendable () async throws -> String)? = nil
  ) throws {
    epoch += 1
    let oldSession = session
    let changed = self.context.userId != context.userId
    self.context = context
    self.identityToken = identityToken
    session = nil
    if changed {
      state = StoredState(userID: context.userId)
      activeThreadID = nil
      publish()
      try persist()
      if let oldSession {
        Task { [api] in
          // Revocation failure is observable, but never restores the old account's UI.
          do { try await api.logout(token: oldSession.token) } catch {
            self.errorMessage =
              "Previous support session could not be revoked. It will expire automatically."
          }
        }
      }
    }
    publish()
  }

  public func refresh() async {
    await operate { epoch in
      var all: [ThreadStatus] = []
      var after: String?
      var seen = Set<String>()
      repeat {
        let page = try await self.authorized(epoch) {
          try await self.api.statuses(token: $0, after: after)
        }
        try self.check(epoch)
        for status in page.threads { _ = try replyCursor(status.latestReplyCursor) }
        all += page.threads
        after = page.nextCursor
        if let after, !seen.insert(after).inserted {
          throw RespondKitError("Repeated history cursor.")
        }
      } while after != nil
      self.state.statuses = all.sorted { $0.thread.updatedAt > $1.thread.updatedAt }
      try self.persist()
      self.publish()
      if self.screenVisible, let id = self.activeThreadID { try await self.loadMessages(id, epoch) }
      try await self.flushReads(epoch)
    }
  }

  public func sendDraft() async {
    let text = draft
    let threadID = activeThreadID
    guard !text.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty, text.utf16.count <= 6_000
    else {
      errorMessage = "Enter a message of up to 6,000 characters."
      return
    }
    await operate { epoch in
      if let threadID,
        self.state.statuses.first(where: { $0.thread.id == threadID })?.thread.state == "closed"
      {
        throw RespondKitError("This conversation is closed. Start a new conversation.")
      }
      let key = threadID ?? "new"
      let pending = PendingMessage(
        id: newID("cmsg"), text: text, acceptedAt: Date().ISO8601Format(), delivery: "sending")
      self.state.pending[key, default: []].append(pending)
      // Preserve edits made while another operation was finishing.
      if self.state.drafts[key] == text { self.state.drafts[key] = "" }
      try self.persist()
      self.publish()
      try await self.deliver(pending, threadID: threadID, epoch: epoch)
    }
  }
  public func retry(_ messageID: String) async {
    let id = activeThreadID
    await operate { epoch in
      guard let pending = self.state.pending[id ?? "new"]?.first(where: { $0.id == messageID })
      else { return }
      try await self.deliver(pending, threadID: id, epoch: epoch)
    }
  }

  /// Called by the SDK screen only after the end of the displayed transcript is visible.
  public func markDisplayed(threadID: String, cursor: String) async {
    guard isForeground, screenVisible, activeThreadID == threadID, loadedCursor == cursor,
      messages.contains(where: \.isReply)
    else { return }
    await operate { epoch in
      guard self.isForeground, self.screenVisible, self.activeThreadID == threadID,
        self.loadedCursor == cursor
      else { return }
      let value = try replyCursor(cursor)
      guard value > (Int64(self.state.readCursors[threadID] ?? "0") ?? 0) else { return }
      self.state.readCursors[threadID] = cursor
      self.state.pendingReads[threadID] = cursor
      try self.persist()
      self.publish()
      try await self.flushReads(epoch)
    }
  }

  private func deliver(_ pending: PendingMessage, threadID: String?, epoch: Int) async throws {
    var id = threadID
    do {
      if id == nil {
        let created = try await authorized(epoch) {
          try await self.api.createThread(token: $0, clientThreadID: self.state.newClientThreadID)
        }
        try check(epoch)
        id = created.id
        state.statuses.removeAll { $0.thread.id == created.id }
        state.statuses.insert(ThreadStatus(thread: created, latestReplyCursor: "0"), at: 0)
        state.pending[created.id, default: []] += state.pending.removeValue(forKey: "new") ?? []
        state.drafts[created.id] = state.drafts.removeValue(forKey: "new")
        state.newClientThreadID = newID("cthread")
        activeThreadID = created.id
        try persist()
        publish()
      }
      guard let id else { return }
      setDelivery(pending.id, threadID: id, value: "sending")
      try persist()
      publish()
      let accepted = try await authorized(epoch) {
        try await self.api.send(
          token: $0, threadID: id, clientMessageID: pending.id, text: pending.text)
      }
      try check(epoch)
      setDelivery(
        pending.id, threadID: id,
        value: accepted.status == "failed"
          ? "failed" : accepted.status == "acceptance_unknown" ? "acceptance_unknown" : "accepted")
      try persist()
      publish()
      try await loadMessages(id, epoch)
    } catch {
      guard epoch == self.epoch else { throw error }
      setDelivery(pending.id, threadID: id ?? "new", value: "acceptance_unknown")
      try persist()
      publish()
      throw error
    }
  }
  private func loadMessages(_ id: String, _ epoch: Int) async throws {
    var cursor = state.cursors[id] ?? "0"
    var ordered = state.messages[id] ?? []
    var more: Bool
    repeat {
      let page = try await authorized(epoch) {
        try await self.api.messages(token: $0, threadID: id, after: cursor)
      }
      try check(epoch)
      let next = try replyCursor(page.nextCursor)
      let previous = try replyCursor(cursor)
      guard next >= previous, !page.hasMore || next > previous else {
        throw RespondKitError("Invalid message pagination.")
      }
      for message in page.messages {
        if let index = ordered.firstIndex(where: { $0.id == message.id }) {
          ordered[index] = message
        } else {
          ordered.append(message)
        }
      }
      cursor = page.nextCursor
      more = page.hasMore
    } while more
    // Preserve transcript order when two messages have the same acceptance timestamp.
    state.messages[id] = ordered.sorted { $0.acceptedAt < $1.acceptedAt }
    state.cursors[id] = cursor
    let canonical = Dictionary(
      (state.messages[id] ?? []).compactMap { message in
        message.clientMessageId.map { ($0, message) }
      }, uniquingKeysWith: { _, new in new })
    state.pending[id] = (state.pending[id] ?? []).compactMap { pending in
      guard let message = canonical[pending.id] else { return pending }
      if message.state != "failed" { return nil }
      var pending = pending
      pending.delivery = "failed"
      return pending
    }
    // A customer message can fail after its optimistic row was reconciled away.
    for message in state.messages[id] ?? [] where message.state == "failed" {
      if let clientID = message.clientMessageId,
        !(state.pending[id] ?? []).contains(where: { $0.id == clientID })
      {
        state.pending[id, default: []].append(
          PendingMessage(
            id: clientID, text: message.text, acceptedAt: message.acceptedAt, delivery: "failed"))
      }
    }
    try persist()
    publish()
  }
  private func flushReads(_ epoch: Int) async throws {
    for (id, cursor) in state.pendingReads {
      try await authorized(epoch) {
        try await self.api.markRead(token: $0, threadID: id, cursor: cursor)
      }
      try check(epoch)
      state.pendingReads.removeValue(forKey: id)
      try persist()
    }
  }
  private func validToken(_ epoch: Int) async throws -> String {
    try persist()
    if let session, let expiry = timestamp(session.expiresAt),
      expiry > Date().addingTimeInterval(15)
    {
      return session.token
    }
    let assertion = try await identityToken?()
    try check(epoch)
    if identityToken != nil && (assertion?.isEmpty ?? true) {
      throw RespondKitError("Support identity could not be verified.")
    }
    let value = try await api.createSession(
      installationID: state.installationID, context: context, identityToken: assertion)
    try check(epoch)
    session = value
    return value.token
  }
  private func authorized<T>(_ epoch: Int, _ action: (String) async throws -> T) async throws -> T {
    do { return try await action(validToken(epoch)) } catch let error as RespondKitError
      where error.status == 401
    {
      try check(epoch)
      session = nil
      return try await action(validToken(epoch))
    }
  }
  private func check(_ expected: Int) throws {
    try Task.checkCancellation()
    guard expected == epoch else { throw CancellationError() }
  }
  private func operate(_ action: (Int) async throws -> Void) async {
    let expected = epoch
    if locked { await withCheckedContinuation { waiters.append($0) } } else { locked = true }
    defer {
      isLoading = false
      if waiters.isEmpty { locked = false } else { waiters.removeFirst().resume() }
    }
    guard expected == epoch, !Task.isCancelled else { return }
    isLoading = true
    do {
      try await action(expected)
      try check(expected)
      errorMessage = nil
    } catch is CancellationError {} catch {
      if expected == epoch { errorMessage = error.localizedDescription }
    }
  }
  private func setDelivery(_ messageID: String, threadID: String, value: String) {
    guard let index = state.pending[threadID]?.firstIndex(where: { $0.id == messageID }) else {
      return
    }
    state.pending[threadID]?[index].delivery = value
  }
  private func persist() throws { try persistence.save(JSONEncoder().encode(state)) }
  private func persistOrReport() {
    do { try persist() } catch { errorMessage = error.localizedDescription }
  }
  private func publish() {
    statuses = state.statuses
    hasUnreadReplies = statuses.contains { isUnread($0.thread.id) }
    let key = activeThreadID ?? "new"
    messages = state.messages[key] ?? []
    pendingMessages = state.pending[key] ?? []
    draft = state.drafts[key] ?? ""
    loadedCursor = state.cursors[key] ?? "0"
  }
}
